### ISSUE TITLE
Is Canary using the right model?

### ISSUE THREAD

**Issue by devshoe** [CopilotC-Nvim/CopilotChat.nvim#271]
Outstanding plugin guys, love the simplicity.

My issue seems to be that with default options, copilot uses GPT-3 rather than 4.  Is there a bug? If I ask it which model its based on it will say GPT 3. Then if I open VSCode and try it out, it responds with GPT 4. 

A simple question I ask it is ''If there are 10 books in a room and I read 2, how many books are still in the room?". GPT 3 always fails and says 8 while 4 responds with 10.
I.E. VSCode Copilot Chat gets it right but Chat.nvim responds with the wrong answer.
 Am I doing something wrong?

![Screenshot 2024-04-11 192934](https://github.com/CopilotC-Nvim/CopilotChat.nvim/assets/43845947/f52a2264-5fdb-4034-8fe1-c4c3514d69c4)

Here is my config:
` 

     
            {
                 "CopilotC-Nvim/CopilotChat.nvim",
            branch = "canary",
                 cmd = {
                         "CopilotChat",
                         "CopilotChatOpen",
                         "CopilotChatClose",
                         "CopilotChatToggle",
                         "CopilotChatExplain",
                         "CopilotChatTests",
                         "CopilotChatFix",                                                                                                 
                             "CopilotChatOptimize",
                            "CopilotChatDocs",
                         "CopilotChatFixDiagnostic",
                           "CopilotChatCommit",                                                                                              
                            "CopilotChatCommitStaged",
                  },
                  dependencies = {
                          { "github/copilot.vim" }, -- or github/copilot.vim                                             
                            { "nvim-lua/plenary.nvim" }, -- for curl, log wrapper
                  },
                  config = function()
                          require("CopilotChat").setup(opts.copilot())
                  end,
         },
`

And my opts.copilot is just the copy pasted base config 

`
```
M.copilot = function()
  1         local prompts = require("CopilotChat.prompts")
  2         local select = require("CopilotChat.select")
  3         return {
  4                 debug = false, -- Enable debug logging
  5                 proxy = nil, -- [protocol://]host[:port] Use this proxy
  6                 allow_insecure = false, -- Allow insecure server connections
  7
  8                 system_prompt = prompts.COPILOT_INSTRUCTIONS, -- System prompt to use
  9                 model = "gpt-4", -- GPT model to use, 'gpt-3.5-turbo' or 'gpt-4'
 10                 temperature = 0.1, -- GPT temperature
 11
 12                 question_header = "## User ", -- Header to use for user questions
 13                 answer_header = "## Copilot ", -- Header to use for AI answers
 14                 error_header = "## Error ", -- Header to use for errors
 15                 separator = "---", -- Separator to use in chat
 16
 17                 show_folds = true, -- Shows folds for sections in chat
 18                 show_help = false, -- Shows help message as virtual lines when waiting for user input
 19                 auto_follow_cursor = true, -- Auto-follow cursor in chat
 20                 auto_insert_mode = false, -- Automatically enter insert mode when opening window and if auto follow curso                    r is enabled on new prompt
 21                 clear_chat_on_new_prompt = false, -- Clears chat on every new prompt
 22
 23                 context = "buffers", -- Default context to use, 'buffers', 'buffer' or none (can be specified manually in                     prompt via @).
 24                 history_path = vim.fn.stdpath("data") .. "/copilotchat_history", -- Default path to stored history
 25                 callback = nil, -- Callback to use when ask response is received
 26
 27                 -- default selection (visual or line)
 28                 selection = function(source)
 29                         return select.visual(source) or select.line(source)
 30                 end,
 31
 32                 -- default prompts
 33                 prompts = {
 34                         Explain = {
 35                                 prompt = "/COPILOT_EXPLAIN Write an explanation for the code above as paragraphs of text.                                    ",
 36                         },
 37                         Tests = {
 38                                 prompt = "/COPILOT_TESTS Write a set of detailed unit test functions for the code above."                                    ,
 39                         },
 40                         Fix = {
 41                                 prompt = "/COPILOT_FIX There is a problem in this code. Rewrite the code to show it with                                     the bug fixed.",
 42                         },
 Optimize = {
 14                                 prompt = "/COPILOT_REFACTOR Optimize the selected code to improve performance and readabl                                    ilty.",
 13                         },
 12                         Docs = {
 11                                 prompt = "/COPILOT_REFACTOR Write documentation for the selected code. The reply should b                                    e a codeblock containing the original code with the documentation added as comments. Use                                     the most appropriate documentation style for the programming language used (e.g. JSDoc fo                                    r JavaScript, docstrings for Python etc.",
 10                         },
  9                         FixDiagnostic = {
  8                                 prompt = "Please assist with the following diagnostic issue in file:",
  7                                 selection = select.diagnostics,
  6                         },
  5                         Commit = {
  4                                 prompt = "Write commit message for the change with commitizen convention. Make sure the t                                    itle has maximum 50 characters and message is wrapped at 72 characters. Wrap the whole me                                    ssage in code block with language gitcommit.",
  3                                 selection = select.gitdiff,
  2                         },
  1                         CommitStaged = {
234                                 prompt = "Write commit message for the change with commitizen convention. Make sure the t                                    itle has maximum 50 characters and message is wrapped at 72 characters. Wrap the whole me                                    ssage in code block with language gitcommit.",
  1                                 selection = function(source)
  2                                         return select.gitdiff(source, true)
  3                                 end,
  4                         },
  5                 },
  6
  7                 -- default window options
  8                 window = {
  9                         layout = "vertical", -- 'vertical', 'horizontal', 'float'
 10                         -- Options below only apply to floating windows
 11                         relative = "editor", -- 'editor', 'win', 'cursor', 'mouse'
 12                         border = "single", -- 'none', single', 'double', 'rounded', 'solid', 'shadow'
 13                         width = 0.8, -- fractional width of parent
 14                         height = 0.6, -- fractional height of parent
 15                         row = nil, -- row position of the window, default is centered
 16                         col = nil, -- column position of the window, default is centered
 17                         title = "Copilot Chat", -- title of chat window
 18                         footer = nil, -- footer of chat window
 19                         zindex = 1, -- determines if window is on top or below other floating windows
 20                 },
 21
 22                 -- default mappings
 23                 mappings = {
 24                         complete = {
 25                                 detail = "Use @<Tab> or /<Tab> for options.",
 26                                 insert = "<Tab>",
 27                         },
 28                         close = {
 29                                 normal = "q",
 16                                 insert = "<C-c>",
 15                         },
 14                         reset = {
 13                                 normal = "<C-l>",
 12                                 insert = "<C-l>",
 11                         },
 10                         submit_prompt = {
  9                                 normal = "<CR>",
  8                                 insert = "<C-s>",
  7                         },
  6                         accept_diff = {
  5                                 normal = "<C-y>",
  4                                 insert = "<C-y>",
  3                         },
  2                         yank_diff = {
  1                                 normal = "gy",
280                         },
  1                         show_diff = {
  2                                 normal = "gd",
  3                         },
  4                         show_system_prompt = {
  5                                 normal = "gp",
  6                         },
  7                         show_user_selection = {
  8                                 normal = "gs",
  9                         },
 10                 },
 11         }
 12 end
 13
 14 return M
```

`

**Comment by deathbeam** [comment]
I just tried the same thing and got this:

![screenshot](https://github.com/CopilotC-Nvim/CopilotChat.nvim/assets/5115805/ae1aa311-4e05-4a53-b874-03ae28214f57)

OpenAI Codex is also based on GPT-3 not 4 at least based on wiki so I think VSCode got that wrong?

**Comment by devshoe** [comment]
Your response just made things click 😂
I am assuming it decides which GPT to use based on prompt rather than the actual config?


I had my workspace set to Neovim config folder when I was debugging this. Quite erratically it would respond with either GPT 3 or 4. I spent 30 mins on it trying to figure it out, went through this codebase trying to debug when exactly it uses GPT4.



If you see the config I pasted above, I have set `context="buffers"`

Now I realize: this mf would say GPT-4 only when I toggled the chat window ON THE BUFFER CONTAINING THE CONFIG.

I.E. it literally read gpt-4 from the config and responded as such. LMAO.
